### PR TITLE
[TOPIC-GPIO]: drivers: gpio_cmsdk_ahb: update to use new GPIO API

### PR DESCRIPTION
**Pull request by galak**
Update driver code and board files to use new GPIO configuration flags
such as GPIO_ACTIVE_LOW. Also add implementation of new port_* driver
API as well as gpio_pin_interrupt_configure function.

We treat GPIO_INT_MODE_LEVEL as not supported as the hardware doesn't
seem to handle level interrupts as one would expect.  Looking into this
further to determine if its a HW bug or some misconfiguration in the
software.

Tested on musca-a board.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

### CODE REVIEW

**Review by mnkp — CHANGES_REQUESTED** [review]

File: drivers/gpio/gpio_cmsdk_ahb.c [hunk 1]
```diff
@@ -34,14 +35,74 @@ struct gpio_cmsdk_ahb_cfg {
...
+
 static int cmsdk_ahb_gpio_config(struct device *dev, u32_t mask, int flags)
 {
 	const struct gpio_cmsdk_ahb_cfg * const cfg = dev->config->config_info;
 
+	if ((flags & GPIO_DIR_MASK) == 0) {
+		return -ENOTSUP;
+	}
+
+	if ((flags & GPIO_PULL_UP) == GPIO_PULL_UP) {
```
[the remark this answers is not in the document]
Comment (galak) [reply drivers/gpio/gpio_cmsdk_ahb.c]: I find this personally harder to read.  I find negative logic not as intuitive as positive logic, but that's my 2 cents.

File: drivers/gpio/gpio_cmsdk_ahb.c [hunk 2]
```diff
@@ -50,6 +111,11 @@ static int cmsdk_ahb_gpio_config(struct device *dev, u32_t mask, int flags)
 	 */
 	if ((flags & GPIO_DIR_MASK) == GPIO_DIR_OUT) {
```
[the remark this answers is not in the document]
Comment (galak) [reply drivers/gpio/gpio_cmsdk_ahb.c]: Fixed.

File: drivers/gpio/gpio_cmsdk_ahb.c [hunk 3]
```diff
@@ -200,6 +266,45 @@ static int gpio_cmsdk_ahb_read(struct device *dev, int access_op,
...
+			cfg->port->inttypeclr = BIT(pin);
+		}
+
+		/* Level High or Edge Risising */
+		if (trig == GPIO_INT_TRIG_HIGH) {
+			cfg->port->intpolset = BIT(pin);
+		} else {
+			cfg->port->intpolclr = BIT(pin);
+		}
+		cfg->port->intenset = BIT(pin);
```
[the remark this answers is not in the document]
Comment (galak) [reply drivers/gpio/gpio_cmsdk_ahb.c]: Changed.

File: drivers/gpio/gpio_cmsdk_ahb.c [hunk 4]
```diff
@@ -50,6 +111,11 @@ static int cmsdk_ahb_gpio_config(struct device *dev, u32_t mask, int flags)
 	 */
 	if ((flags & GPIO_DIR_MASK) == GPIO_DIR_OUT) {
 		cfg->port->outenableset = mask;
```
[the remark this answers is not in the document]
Comment (galak) [reply drivers/gpio/gpio_cmsdk_ahb.c]: Re-ordered to address issue.

File: drivers/gpio/gpio_cmsdk_ahb.c [hunk 5]
```diff
@@ -40,17 +41,80 @@ struct gpio_cmsdk_ahb_dev_data {
...
+	}
+
 	/*
 	 * Setup the pin direction
 	 * Output Enable:
 	 * 0 - Input
 	 * 1 - Output
 	 */
-	if ((flags & GPIO_DIR_MASK) == GPIO_DIR_OUT) {
+	if (flags & GPIO_OUTPUT) {
```
Comment (mnkp) [inline drivers/gpio/gpio_cmsdk_ahb.c]: We try to be Misra-C compliant elsewhere in the driver code so it's better to write this as
```
	if ((flags & GPIO_OUTPUT) != 0) {
```

File: drivers/gpio/gpio_cmsdk_ahb.c [hunk 6]
```diff
@@ -40,17 +41,80 @@ struct gpio_cmsdk_ahb_dev_data {
...
+		return -ENOTSUP;
+	}
+
+	if ((flags & GPIO_PULL_UP) != 0) {
+		return -ENOTSUP;
+	}
+
+	if ((flags & GPIO_PULL_DOWN) != 0) {
+		return -ENOTSUP;
+	}
```
Comment (mnkp) [inline drivers/gpio/gpio_cmsdk_ahb.c]: We can write it as
```
	if ((flags & (GPIO_PULL_UP | GPIO_PULL_DOWN)) != 0) {
￼		return -ENOTSUP;
￼	}
```
this should generate smaller code size (unless this is already fixed by the compiler optimization step).

**Review by mnkp — CHANGES_REQUESTED** [review]
Looks good, however there is some old interrupt configuration code left over. In the `cmsdk_ahb_gpio_config` function we need to remove the whole `if (flags & GPIO_INT) {` statement.

File: drivers/gpio/gpio_cmsdk_ahb.c [hunk 7]
```diff
@@ -40,17 +41,80 @@ struct gpio_cmsdk_ahb_dev_data {
...
+
```
Comment (mnkp) [inline drivers/gpio/gpio_cmsdk_ahb.c]: We need to add one more check
```
	if ((flags & GPIO_SINGLE_ENDED) != 0) {
		return -ENOTSUP;
	}
```

Comment (galak) [reply drivers/gpio/gpio_cmsdk_ahb.c]: added

File: drivers/gpio/gpio_cmsdk_ahb.c (hunk 6 above)
Comment (galak) [reply drivers/gpio/gpio_cmsdk_ahb.c]: Changed.

File: drivers/gpio/gpio_cmsdk_ahb.c (hunk 5 above)
Comment (galak) [reply drivers/gpio/gpio_cmsdk_ahb.c]: Changed.

**Review by mnkp — CHANGES_REQUESTED** [review]
There is still some old interrupt configuration code left over in the `cmsdk_ahb_gpio_config` function. Lines 124 to 149.

**Comment by galak** [comment]
> There is still some old interrupt configuration code left over in the `cmsdk_ahb_gpio_config` function. Lines 124 to 149.

removed.

**Review by mnkp — APPROVED** [review]

**Pull request merged**